torch/onnx/symbolic.py

```diff
@@ -1090,3 +884,133 @@ def symbolic_flattened_wrapper(g, input, *args):
+            return _unimplemented("RNN/GRU/LSTM", "dropout in training mode")
+
+        unidirectional = not bidirectional
+
+        prev_output = input
+
+        if variant == 'RNN' or variant == 'GRU':
+            h0 = initial_states
+            h_outs = []
+        elif variant == 'LSTM':
+            h0, c0 = initial_states
+            h_outs = []
+            c_outs = []
+
+        sequence_lens = unused(g) if batch_sizes is None else batch_sizes
+
+        for i in range(num_layers):
+            if unidirectional:
+                if variant == 'RNN':
+                    weight_ih, weight_hh, bias_ih, bias_hh = all_weights[i]
+                elif variant == 'GRU':
+                    # pytorch is reset, input, hidden
+                    # onnx is    input, reset, hidden
+                    weight_ih, weight_hh, bias_ih, bias_hh = \
+                        [reform_weights(g, w, hidden_size, [(1, 2), (0, 1), (2, 3)]) for w in all_weights[i]]
+                elif variant == 'LSTM':
+                    # pytorch is input, forget, cell, output.
+                    # onnx is    input, output, forget, cell.
+                    weight_ih, weight_hh, bias_ih, bias_hh = \
+                        [reform_weights(g, w, hidden_size, [(0, 1), (3, 4), (1, 3)]) for w in all_weights[i]]
+                bias_concat = g.op('Concat', bias_ih, bias_hh, axis_i=0)
+
+                # Unidirectional weights in PyTorch omit the
+                # directionality dimension (dim=0), whereas in ONNX it
+                # is required even for unidirectional RNN; thus,
+                # insert a dimension of size at dim=0.
+                weight_ih = g.op('Unsqueeze', weight_ih, axes_i=[0])
+                weight_hh = g.op('Unsqueeze', weight_hh, axes_i=[0])
+                bias_concat = g.op('Unsqueeze', bias_concat, axes_i=[0])
+
+                if variant == 'RNN' or variant == 'GRU':
+                    state_inputs = [
+                        h0 if num_layers == 1 else g.op('Slice', h0, axes_i=[0], starts_i=[i], ends_i=[i + 1])
+                    ]
+                elif variant == 'LSTM':
+                    state_inputs = [
+                        h0 if num_layers == 1 else g.op('Slice', h0, axes_i=[0], starts_i=[i], ends_i=[i + 1]),
+                        c0 if num_layers == 1 else g.op('Slice', c0, axes_i=[0], starts_i=[i], ends_i=[i + 1])
+                    ]
+            else:
+                if variant == 'RNN':
+                    weight_ih_f, weight_hh_f, bias_ih_f, bias_hh_f = all_weights[2 * i]
+                    weight_ih_b, weight_hh_b, bias_ih_b, bias_hh_b = all_weights[2 * i + 1]
+                elif variant == 'GRU':
+                    # pytorch is reset, input, hidden
+                    # onnx is    input, reset, hidden
+                    weight_ih_f, weight_hh_f, bias_ih_f, bias_hh_f = \
+                        [reform_weights(g, w, hidden_size, [(1, 2), (0, 1), (2, 3)]) for w in all_weights[2 * i]]
```

Comment:
Can you deduplicate the permutation here? It's duplicated with the `reform_weights` invocation above. Even better would also be to avoid the top-level if all together by applying identity intervals in the RNN case. So then it looks like:

```
if variant == 'RNN':
  weight_intervals = [(0,1), (1,2), (2,3)]
elif variant == 'GRU':
  weight_intervals = [(1, 2), (0, 1), (2, 3)]
elif variant == 'LSTM':
  weight_intervals = [(0, 1), (3, 4), (1, 3)]

# ...

                    weight_ih_f, weight_hh_f, bias_ih_f, bias_hh_f = \
                        [reform_weights(g, w, hidden_size, weight_permutation) for w in all_weights[2 * i]]
                    weight_ih_b, weight_hh_b, bias_ih_b, bias_hh_b = \
                        [reform_weights(g, w, hidden_size, weight_permutation) for w in all_weights[2 * i + 1]]
```

Reply:
well, this does generate a ton of operators in the resulting graph, so I would like to skip it entirely in the RNN case (but I can add a comment saying this). Otherwise yes should be possible